### PR TITLE
feat[next]: Add missing UnitRange comparison functions

### PR DESCRIPTION
**Pull request by havogt**

### CODE REVIEW

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/gt4py/next/common.py [hunk 1]
```diff
@@ -141,15 +141,68 @@ def __and__(self, other: Set[int]) -> UnitRange:
         else:
             raise NotImplementedError("Can only find the intersection between UnitRange instances.")
 
-    def __le__(self, other: Set[int]):
+    def __contains__(self, value: Any) -> bool:
+        if not isinstance(value, core_defs.INTEGRAL_TYPES):
+            return False
+        if value == Infinity.positive() or value == Infinity.negative():
+            # raising error was an adhoc decision, feel free to improve
+            raise ValueError("Cannot check if Infinity is in UnitRange.")
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:149]: Ok, let's talk

File: tests/next_tests/unit_tests/test_common.py [hunk 2]
```diff
@@ -35,6 +35,11 @@
 KDim = Dimension("KDim", kind=DimensionKind.VERTICAL)
 
 
+@pytest.fixture(params=[Infinity.positive(), Infinity.negative()])
```
[the remark this answers is not in the document]
Comment (havogt) [reply tests/next_tests/unit_tests/test_common.py]: Now it's used in 2 places. ;)

File: src/gt4py/next/common.py [hunk 3]
```diff
@@ -84,32 +73,113 @@ def __str__(self):
         return f"{self.value}[{self.kind}]"
 
 
+@dataclasses.dataclass(frozen=True)
+class _Infinity:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:77]: Look at `__and__` and `__add__`, they are trivial now. Rewrite `__and__` without the helper, it's super ugly. Basically I started with what we discussed and then it became very ugly.

File: src/gt4py/next/common.py [hunk 4]
```diff
@@ -84,32 +73,113 @@ def __str__(self):
...
+    stop: int | _Infinity
 
-    def __init__(self, start: core_defs.IntegralScalar, stop: core_defs.IntegralScalar) -> None:
-        if start < stop:
-            object.__setattr__(self, "start", int(start))
-            object.__setattr__(self, "stop", int(stop))
+    def __init__(
+        self,
+        start: core_defs.IntegralScalar | _Infinity | None,
+        stop: core_defs.IntegralScalar | _Infinity | None,
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:125]: Yes, here the abstraction leaks, so after we concluded if Infinity make sense as implementation detail, we need to figure out how to remove it from the public constructor.

File: src/gt4py/next/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py:77]: Ok, yes, I was missing the point that the sign attribute (`pos`) is needed for some comparisons. I would propose then to define an Enum with two sentinel values for the open lower and upper bounds and customize the comparison operators as in the [OrderedEnum](https://devdocs.io/python~3.11/howto/enum#orderedenum) example:

```python
class _Unbounded(enum.Enum):
    LOWER = enum.auto()
    UPPER = enum.auto()

    def __add__(self, _):
        return self

    __radd__ = __add__

    def __sub__(self, _):
        return self

    __rsub__ = __sub__

    def __le__(self, other):
        return self is self.LOWER or other is self.UPPER

    def __lt__(self, other):
        return self is self.LOWER and other is not self

    def __ge__(self, other):
        return self is self.UPPER or other is self.LOWER

    def __gt__(self, other):
        return self is self.UPPER and other is not self

```

[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:77]: thanks! that's useful

File: src/gt4py/next/common.py [hunk 5]
```diff
@@ -84,32 +73,113 @@ def __str__(self):
...
+    def unbound(cls) -> UnitRange:
+        return cls(None, None)
+
+    @property
+    def bound_start(self) -> int:
+        assert not isinstance(self.start, _Infinity)
+        return self.start
+
+    @property
+    def bound_stop(self) -> int:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:165]: done

File: src/gt4py/next/common.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: made it public

File: src/gt4py/next/common.py [hunk 6]
```diff
@@ -84,32 +73,113 @@ def __str__(self):
...
 
+    def is_left_unbound(self):
+        return isinstance(self.start, _Infinity)
+
+    def is_right_unbound(self):
+        return isinstance(self.stop, _Infinity)
+
+    def is_unbound(self):
+        return self.is_left_unbound() or self.is_right_unbound()
+
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: using a mix of open and bounded now

File: src/gt4py/next/common.py [hunk 7]
```diff
@@ -140,48 +210,82 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:  # noqa: F811 # re
...
+            return Set.__lt__(self, other)
+
+    def __ge__(self, other: Set[int]) -> bool:
+        if isinstance(other, UnitRange):
+            return self.start <= other.start and self.stop >= other.stop
+        elif self.is_unbound():
+            for v in other:
+                if v not in self:
+                    return False
+            return True
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:240]: done

File: src/gt4py/next/common.py [hunk 8]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
         return f"{self.value}[{self.kind}]"
 
 
+class OpenBound(enum.Enum):
```
Comment (egparedes) [inline src/gt4py/next/common.py:77]: Missing doctring

File: src/gt4py/next/common.py [hunk 9]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
         else:
             # make UnitRange(0,0) the single empty UnitRange
             object.__setattr__(self, "start", 0)
             object.__setattr__(self, "stop", 0)
 
-    # TODO: the whole infinity idea and implementation is broken and should be replaced
     @classmethod
-    def infinity(cls) -> UnitRange:
-        return cls(Infinity.negative(), Infinity.positive())
+    def unbound(cls) -> UnitRange:
```
Comment (egparedes) [inline src/gt4py/next/common.py:149]:
```suggestion
    def unbounded(cls) -> UnitRange:
```

File: src/gt4py/next/common.py [hunk 10]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
     def __len__(self) -> int:
-        if Infinity.positive() in (abs(self.start), abs(self.stop)):
-            return Infinity.positive()
-        return max(0, self.stop - self.start)
+        if self.is_left_open() or self.is_right_open():
+            raise ValueError("Cannot compute length of open UnitRange.")
+        return max(0, self.bounded_stop - self.bounded_start)
+
+    def is_left_open(self):
+        return isinstance(self.start, OpenBound)
```
Comment (egparedes) [inline src/gt4py/next/common.py:168]:
```suggestion
        return self.start is OpenBound.LOWER
```

File: src/gt4py/next/common.py [hunk 11]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
+
+    def is_right_open(self):
+        return isinstance(self.stop, OpenBound)
```
Comment (egparedes) [inline src/gt4py/next/common.py:171]:
```suggestion
        return self.stop is OpenBound.UPPER
```

File: src/gt4py/next/common.py [hunk 12]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
+        start: core_defs.IntegralScalar | OpenBound | None,
+        stop: core_defs.IntegralScalar | OpenBound | None,
+    ) -> None:
+        if (
+            start is None
+            or isinstance(start, OpenBound)
+            or stop is None
+            or isinstance(stop, OpenBound)
+            or start < stop
+        ):
```
Comment (egparedes) [inline src/gt4py/next/common.py:126]:
```suggestion
        if (
            start in (None, OpenBound.LOWER)
            or stop in (None, OpenBound.UPPER)
            or start < stop
        ):
```

File: src/gt4py/next/common.py [hunk 13]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
+            if start is None:
+                start = OpenBound.LOWER
+            elif isinstance(start, OpenBound):
+                assert start is OpenBound.LOWER
```
Comment (egparedes) [inline src/gt4py/next/common.py:130]: Combined with the suggestion above, I don't think we need to assert anything here. In my opinion the case where the wrong `OpenBound` is passed to the start/stop arg should be treated exactly the same as the case where `start > stop` in the outer `else` branch:
```suggestion
            elif start is OpenBound.LOWER:
```

File: src/gt4py/next/common.py [hunk 14]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
+                start = start
+            else:
+                start = int(start)
+            if stop is None:
+                stop = OpenBound.UPPER
+            elif isinstance(stop, OpenBound):
+                assert stop is OpenBound.UPPER
```
Comment (egparedes) [inline src/gt4py/next/common.py:137]: Same comment as above:
```suggestion
            elif stop is OpenBound:
```

File: src/gt4py/next/common.py [hunk 15]
```diff
@@ -84,31 +74,104 @@ def __str__(self):
...
 
-    def __init__(self, start: core_defs.IntegralScalar, stop: core_defs.IntegralScalar) -> None:
-        if start < stop:
-            object.__setattr__(self, "start", int(start))
-            object.__setattr__(self, "stop", int(stop))
+    def __init__(
+        self,
+        start: core_defs.IntegralScalar | OpenBound | None,
+        stop: core_defs.IntegralScalar | OpenBound | None,
+    ) -> None:
```
Comment (egparedes) [inline src/gt4py/next/common.py:119]: Just a comment: it seems that `None` values are supported in the initializer args just as a convenient way for users to specify open ranges, but it makes the implementation slightly more complex than strictly needed. An alternative could be to create two builder methods for the cases where the lower or upper bounds are open, in the same style as the `unbounded()` constructor. For example:
```python
@classmethod
def stop_at(cls, stop: int) -> UnitRange:
    return UnitRange(OpenBound.LOWER, stop)

@classmethod
def start_at(cls, start int) -> UnitRange:
    return UnitRange(start, OpenBound.UPPER)

```

File: tests/next_tests/unit_tests/test_common.py [hunk 16]
```diff
@@ -41,6 +41,54 @@ def a_domain():
...
+    assert op(value, OpenBound.LOWER)
+
+    assert not op(OpenBound.LOWER, OpenBound.UPPER)
+
+
+def test_unbounded_eq(unbounded):
+    assert unbounded == unbounded
+    assert unbounded <= unbounded
+    assert unbounded >= unbounded
+
```
Comment (egparedes) [inline tests/next_tests/unit_tests/test_common.py:82]:
```suggestion
assert not (unbounded < unbounded)
assert not (unbounded > unbounded)
```

**Review by egparedes — COMMENTED** [review]
I agree with the changes but I have a few more suggestions and comments.

File: src/gt4py/next/common.py [hunk 17]
```diff
@@ -141,15 +141,68 @@ def __and__(self, other: Set[int]) -> UnitRange:
...
-    def __le__(self, other: Set[int]):
+    def __contains__(self, value: Any) -> bool:
+        if not isinstance(value, core_defs.INTEGRAL_TYPES):
+            return False
+        if value == Infinity.positive() or value == Infinity.negative():
+            # raising error was an adhoc decision, feel free to improve
+            raise ValueError("Cannot check if Infinity is in UnitRange.")
+        return value >= self.start and value < self.stop
+
+    def __le__(self, other: Set[int]) -> bool:
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py]: do you still want what you proposed yesterday, now that the functions are much simpler?

File: src/gt4py/next/common.py (hunk 9 above)
Comment (havogt) [reply src/gt4py/next/common.py:149]: -> open

File: src/gt4py/next/common.py (hunk 13 above)
Comment (havogt) [reply src/gt4py/next/common.py:130]: agree

File: src/gt4py/next/common.py (hunk 15 above)
Comment (havogt) [reply src/gt4py/next/common.py:119]: After you suggested changes, it's still slightly more complex than needed, but also relatively straight-forward. Do you still think we need the builders?

File: src/gt4py/next/common.py (hunk 9 above)
Comment (havogt) [reply src/gt4py/next/common.py:149]: flake8 complains about shadowing a builtin `open`, but I don't see why a classmethod shadows a free function, what do I miss?

File: src/gt4py/next/embedded/common.py [hunk 18]
```diff
@@ -52,8 +52,8 @@ def _relative_sub_domain(
         else:
             # not in new domain
             assert common.is_int_index(idx)
-            new_index = (rng.start if idx >= 0 else rng.stop) + idx
-            if new_index < rng.start or new_index >= rng.stop:
+            new_index = (rng.bounded_start if idx >= 0 else rng.bounded_stop) + idx
+            if new_index < rng.bounded_start or new_index >= rng.bounded_stop:
```
Comment (tehrengruber) [inline src/gt4py/next/embedded/common.py:56]: So you use `bounded_start`, but don't handle the unbound case?

**Review by tehrengruber — CHANGES_REQUESTED** [review]

File: src/gt4py/next/common.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/common.py]: I won't push for it anymore, but see my comment why I still like it for some of them.

**Review by egparedes — APPROVED** [review]
I have a few style comments but overall looks good so I'm approving the PR because the comments are optional.

File: src/gt4py/next/common.py [hunk 19]
```diff
@@ -283,18 +341,21 @@ def named_range(v: tuple[Dimension, RangeLike]) -> NamedRange:
     return (v[0], unit_range(v[1]))
 
 
+_Rng = TypeVar("_Rng", bound=UnitRange)
```
[the remark this answers is not in the document]
Comment (havogt) [reply src/gt4py/next/common.py:344]: maybe
```
_Rng = TypeVar(
    "_Rng",
    UnitRange[int, int],
    UnitRange[Infinity, int],
    UnitRange[int, Infinity],
    UnitRange[Infinity, Infinity],
)
```

**Comment by havogt** [comment]
@tehrengruber I will merge this PR. If you have more comments we can discuss them in a separate PR.

**Pull request merged**